app/services/custom_fields/custom_field_projects/delete_service.rb

```diff
@@ -0,0 +1,53 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module CustomFields
+  module CustomFieldProjects
+    class DeleteService < ::BaseServices::Delete
+      def destroy(custom_field_project)
+        delete_result = delete(custom_field_id: custom_field_project.custom_field_id,
+                               project_id: custom_field_project.project_id)
+        ActiveRecord::Type::Boolean.new.cast(delete_result)
+      end
+
+      # `custom_fields_projects` table has no `id` column, hence no primary key. #destroy method would not work
+      # Note: `delete_all` goes straight to the database and does not trigger callbacks
+      #
+      # @return [Integer] number of rows deleted
+      def delete(custom_field_id:, project_id:)
+        CustomFieldsProject.transaction do
+          CustomFieldsProject.where(custom_field_id:, project_id:).delete_all
+        end
+      end
```

Comment:
\cc @ulferts any thoughts on this approach?

Reply:
I think it is ok to do this given what the data structure currently offers. 

But there might be a better way of doing it. I don't have first hand experience with it yet, though, so I don't particularly know the limitations. Since [Rails 7.1.2 composite primary keys are supported](https://blog.saeloun.com/2024/08/06/active-record-composite-primary-key/). I still don't quite understand how much of [this](https://guides.rubyonrails.org/active_record_composite_primary_keys.html) is only available in edge and how much is already in 7.1.2. so you'd have to look that up for yourself. 

But my quick test where I changed the `CustomFieldsProject` to include the specification for the composite primary key

```
class CustomFieldsProject < ApplicationRecord
  self.primary_key = %i[project_id custom_field_id]

  belongs_to :project
  belongs_to :custom_field
end
```

Let me to already do this in the console:

```
[3] pry(main)> CustomFieldsProject.find([992, 152]).destroy
  CustomFieldsProject Load (0.4ms)  SELECT "custom_fields_projects".* FROM "custom_fields_projects" WHERE "custom_fields_projects"."project_id" = $1 AND "custom_fields_projects"."custom_field_id" = $2 LIMIT $3  [["project_id", 992], ["custom_field_id", 152], ["LIMIT", 1]]
  TRANSACTION (0.1ms)  BEGIN
  CustomFieldsProject Destroy (0.7ms)  DELETE FROM "custom_fields_projects" WHERE "custom_fields_projects"."project_id" = $1 AND "custom_fields_projects"."custom_field_id" = $2  [["project_id", 992], ["custom_field_id", 152]]
  TRANSACTION (2.2ms)  COMMIT
=> #<CustomFieldsProject:0x0000000332b9e620 custom_field_id: 152, project_id: 992>
```

I can now also query for e.g. `CustomFieldsProject.last`. It still isn't a proper composite key in the database but it might well be good enough.

If that doesn't work, it might still be worthwhile to add an id column to the table and work through the necessary changes. It will probably make a whole lot of code simpler (e.g. in the controller). But I'd understand if you don't want to do that now.

Reply:
Thank you! In the end adding an ID column seems like the better solution, albeit with higher opportunity cost compared to the composite primary keys which will have the carrying cost of a different data structure- which I have already hit issues with in one part of the application (ng auto-completer)